Improve usability of run loading on fitting tab of Engineering Diffraction UI

DannyHindson opened the pull request:
**Description of work.**

The File Finder widget on the Fitting tab has been updated to filter the files shown in the Browse dialog based on two combo boxes. The combo boxes allow the user to specify a unit (TOF or dSpacing) and\or a grouping (basically North or South bank):

![image](https://user-images.githubusercontent.com/56295817/126459236-d49a3690-5553-423c-8921-dbcf0662f013.png)

In order to get this to work a switch has been added to the FileFinderWidget that is fed through to the underlying QFileDialog object. QFileDialog can be set to use either a Qt file browse dialog or the OS-native file dialog. The Qt version is easier to customise and you can add custom file filters to it. On Windows at least it looks a bit more dated than the OS native dialog but hopefully overall this is an improvement. The switch is a method called FileFinderWidget::setUseNativeWidget

Note - there was already a unit dropdown prior to this PR. It used to drive a unit conversion into whatever unit was selected. This has been repurposed as a file filter and the unit conversion has been removed. The unit conversion feature is a hangover from an earlier version of the Focus tab where it used to output files in TOF only. The focussing now outputs TOF and dSpacing data so there's no longer a need for any unit conversion to occur

**To test:**

It's easiest to test if you have some focused data files already generated for a few runs. You'll need ISIS data archive access:

1. Go to Calibration tab
2. Fill in vanadium 236516 ceria 193749 
3. Hit Calibrate
4. Go to Focus tab
5. Enter 299080-299084
6. Hit Focus

Having done this go to the Fitting tab
Hit Browse
Based on an earlier PR (#32016) the folder should default to the folder where the focussing outputs the files. If for any reason this isn't what you see, go to the Engineering_Mantid\Focus folder in your user area (~ on linux or `C:\Users\<fedid>\ `on Windows)
You should see the set of focused .nxs files you just generated:

![image](https://user-images.githubusercontent.com/56295817/126462178-df307957-64d5-4d18-9026-bddced079c7f.png)

Observe that you only see a subset of file based on the choices made on the TOF or Grouping combo boxes

Finally, worth checking I've not broken the file finder widget by going to another user interface that uses eg Indirect, Corrections UI

Fixes #31690. 


---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DannyHindson committed:
Initial changes to support better file filters on Fitting tab

Add some methods to the FileFinderWidget to support using the
non-native widget
Use the non-native file browser widget in the fitting tab of the
engineering diffraction user interface because it supports more
sophisticated file filters eg based on units and\or bank

DannyHindson committed:
Tidy up non-native logic in FileFinder

DannyHindson committed:
Further tidy up on new file filters

Remember the last directory by storing in ini file
Move the filter dropdowns onto their own row on Fitting tab
Fix error on reopening UI due to FileFilterProxyModel
Remove redundant new folder icon from non-native widget
Update text on bank combo to align with Calibration tab and to
replace blank entry with All

DannyHindson committed:
Add release note and fix up tests

Removal of the unit conversion functionality from Fitting tab
had caused various test failures

DannyHindson committed:
Remove commented out code

DannyHindson commented:
@RichardWaiteSTFC - one for when you're back from holiday. Please can you have a look at this PR about the file loading on the fitting tab. I've labelled it as draft at the moment because ahead of a full PR review I was interested in your thoughts about:

a) is the look\feel of the solution OK? In particular I've used a switch inside the to the QFileDialog which uses the Qt widget rather than the native OS one. On Windows at least the Qt dialog is a bit more clunky\dated but perhaps overall it's an improvement given it allows the ability to add the file filters

b) I've repurposed the Unit combo box to drive the file filtering. As part of this I've removed the ConvertUnits call that was done based on this combo box. There were some comments in the code about this change. One of them talked about being able to read the unit from the x axis of the workspace and calling RenameWorkspace -. There seems to be some code in the fitting model that reads the unit from the x axis already. I'm not doing the part about calling RenameWorkspace at the moment so not sure if I done this correctly

DannyHindson committed:
Fix failing unit test due to unit suffix

RichardWaiteSTFC reviewed (commented):
Works well, I think it's a big improvement being able to filter the files and that the slightly retro file finder is fit for purpose. After PR #32016 the file browser is automatically populated with the TOF files last focussed so even if the file filter is d-spacing it will load TOF - I think this behaviour should be obvious to the users (and anyway TOF is the unit most users will want to load) but I thought it worth mentioning as it might be a slight pain for users to select the d-spacing files for the same run numbers. If you think it is an issue I see two options

1.  Add run start and end filter and populate on focussing (but I think we want to reduce the number of boxes on an already busy interface).
2. Try manipulating the strings to have the unit prefix in the file filter when the file browser box is updated on focussing (but then it is more tightly bound tot he format of the filenames). 

I did have a few other comments:

- I think your code tries to take the file finder to the focus folder in the directory in the settings, but that isn't what I found (although post focussing some new data it does seem to remember my directory).
- There are no specific labels on the unit of crop combo boxes - it's not a big deal and it is fairly obvious but maybe some tool tips?
- You can fool the file finder by adding a TOF (uppercase, it's case sensitive) to the filename of a file, I think the only way to fix this is to be stricter about the filename format (e.g. look for `_TOF.nxs` - but maybe we want to avoid hard-coding the filename?

DannyHindson commented on scripts/Engineering/gui/engineering_diffraction/tabs/fitting/data_handling/data_view.py:
```diff
@@ -8,34 +8,63 @@
         super(FittingDataView, self).__init__(parent)
         self.setupUi(self)
         # file finder
+        self.finder_data.readSettings(path_handling.INTERFACES_SETTINGS_GROUP + '/' + path_handling.ENGINEERING_PREFIX)
```
This change just makes the browse dialog on the Fitting tab remember the folder that the user last selected a file from. So it will only return to the focussed files folder if you've previously selected a file from there
Note that if a path has been automatically filled in by the user having run Focus just before then this takes precedence over the path read from the settings and it will also gets saved as the last folder when you shut down the UI.

DannyHindson commented:
> Works well, I think it's a big improvement being able to filter the files and that the slightly retro file finder is fit for purpose. After PR #32016 the file browser is automatically populated with the TOF files last focussed so even if the file filter is d-spacing it will load TOF - I think this behaviour should be obvious to the users (and anyway TOF is the unit most users will want to load) but I thought it worth mentioning as it might be a slight pain for users to select the d-spacing files for the same run numbers. If you think it is an issue I see two options
> 
> 1. Add run start and end filter and populate on focussing (but I think we want to reduce the number of boxes on an already busy interface).
> 2. Try manipulating the strings to have the unit prefix in the file filter when the file browser box is updated on focussing (but then it is more tightly bound tot he format of the filenames).

I think that trying to get the filters to interact with the path in the file finder edit box could get quite complicated. Not necessarily impossible but I think I'd prefer to keep the filters so they just affect what you see in the Browse dialog for now. I could perhaps emphasize this by changing the label on the filter combo boxes to "Browse filters" instead of "File filters"
> 
> I did have a few other comments:
> 
> * I think your code tries to take the file finder to the focus folder in the directory in the settings, but that isn't what I found (although post focussing some new data it does seem to remember my directory).

As I mentioned in my other reply, the code was meant to remember the last folder where a file was selected. So it relies on the user finding the focus folder on the first use of the UI and it then remembers it

> * There are no specific labels on the unit of crop combo boxes - it's not a big deal and it is fairly obvious but maybe some tool tips?

Good idea, I'll add these
> * You can fool the file finder by adding a TOF (uppercase, it's case sensitive) to the filename of a file, I think the only way to fix this is to be stricter about the filename format (e.g. look for `_TOF.nxs` - but maybe we want to avoid hard-coding the filename?

I can tighten this up and also make it case insensitive. The file name format is already baked into the filtering to some extent so not sure this makes it especially worse

DannyHindson committed:
Tidy up of new filters on Fitting tab

Add All item to units filter so user can load files with any filename
Make filter case insensitive and tighten filter to unit specified must
be at the end of the filename
Change label to Browse filters so more obvious only applies to Browse
dialog

DannyHindson committed:
Adjust path_handling import following folder move in master

DannyHindson committed:
Fix initialisation order compiler warning

DannyHindson committed:
Stop native file dialog returning file when Cancel pressed

RichardWaiteSTFC commented:
Looks good and the file filter is a bit more robust now thanks for that! One comment is that perhaps we could make the xunit filter TOF by default? This is the most common use case and also it then makes it clearer to user that the other box is for the ROI (only other choice user has to make after xunit) - I know you added tool tips (thanks!) which helps but otherwise the user has two boxes saying 'All' and it's not clear what they correspond to.

DannyHindson commented:
> Looks good and the file filter is a bit more robust now thanks for that! One comment is that perhaps we could make the xunit filter TOF by default? This is the most common use case and also it then makes it clearer to user that the other box is for the ROI (only other choice user has to make after xunit) - I know you added tool tips (thanks!) which helps but otherwise the user has two boxes saying 'All' and it's not clear what they correspond to.

Yes no problem. I'll change that shortly

DannyHindson committed:
Make TOF the default value on Units filter combo

RichardWaiteSTFC reviewed (dismissed):
Great, all works nicely and thanks for changing the default xunit

DannyHindson committed:
Fix failing test

Fix problem due to conflict with recently merged change to path_handling.py

RichardWaiteSTFC reviewed (approved):
Looks good now - test is passing
